src/mem/metaslab.h

```diff
@@ -179,6 +208,189 @@ namespace snmalloc
+       */
+      Two
+    };
+
+    /**
+     * Proxy class that allows setting and reading back the bits in each word
+     * that are exposed for the back end.
+     *
+     * The back end must not keep instances of this class after returning the
+     * corresponding meta entry to the front end.
+     */
+    class BackendStateWord
+    {
+      /**
+       * A pointer to the relevant word.
+       */
+      uintptr_t* val;
+
+    public:
+      /**
+       * Constructor, wraps a `uintptr_t`.  Note that this may be used outside
+       * of the meta entry by code wishing to provide uniform storage to things
+       * that are either in a meta entry or elsewhere.
+       */
+      constexpr BackendStateWord(uintptr_t* v) : val(v) {}
+
+      /**
+       * Copy constructor.  Aliases the underlying storage.  Note that this is
+       * not thread safe: two `BackendStateWord` instances sharing access to the
+       * same storage must not be used from different threads without explicit
+       * synchronisation.
+       */
+      constexpr BackendStateWord(const BackendStateWord& other) = default;
+
+      /**
+       * Read the value.  This zeroes any bits in the underlying storage that
+       * the back end is not permitted to access.
+       */
+      uintptr_t get() const
+      {
+        return (*val) & ~BackendReservedMask;
+      }
+
+      /**
+       * Default copy assignment.  See the copy constructor for constraints on
+       * using this.
+       */
+      BackendStateWord& operator=(const BackendStateWord& other) = default;
+
+      /**
+       * Assignment operator.  Zeroes the bits in the provided value that the
+       * back end is not permitted to use and then stores the result in the
+       * value that this class manages.
+       */
+      BackendStateWord& operator=(uintptr_t v)
+      {
+        SNMALLOC_ASSERT_MSG(
+          ((v & BackendReservedMask) == 0),
+          "The back end is not permitted to use the low bits in the meta "
+          "entry. ({} & {}) == {}.",
+          v,
+          BackendReservedMask,
+          (v & BackendReservedMask));
+        *val = v | (*val & BackendReservedMask);
+        return *this;
+      }
+
+      /**
+       * Comparison operator.  Performs address comparison *not* value
+       * comparison.
+       */
+      bool operator!=(const BackendStateWord& other) const
+      {
+        return val != other.val;
+      }
+
+      /**
+       * Returns the address of the underlying storage in a form that can be
+       * passed to `snmalloc::message` for printing.
+       */
+      address_t printable_address()
+      {
+        return address_cast(val);
+      }
+    };
+
+    /**
+     * Get a proxy that allows the back end to read from and write to (some bits
+     * of) a word in the meta entry.  The meta entry must either be unowned or
+     * explicitly claimed by the back end before calling this.
+     */
+    BackendStateWord get_backend_word(Word w)
```

Comment:
Is it worth moving this to `template`-d on `w`, since the only callers know statically which they use?

Reply:
I made it dynamic because it's data-dependent in the caller currently and it could be simplified to a single call.  If it's inlined (which it should be, since it's tiny), the caller can then inline it once and the compiler will find it easier to generate better code than if it has to do ICF.